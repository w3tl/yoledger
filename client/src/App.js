/* eslint-disable react/jsx-filename-extension */
import React, { Component } from 'react';
import { BrowserRouter as Router } from 'react-router-dom';
import { ApolloProvider } from 'react-apollo';
import decode from 'jwt-decode';
import './App.css';
import Navigation from './components/Navigation';
import Routes from './components/Routes';
import apolloClient from './apollo';

export function isLoggedIn() {
  const token = localStorage.getItem('token');
  if (token) {
    const { exp } = decode(token);
    return exp >= Date.now() / 1000;
  }
  return !!token;
}
// TODO: remove token when expires
class App extends Component {
  state = { loggedIn: isLoggedIn(), client: apolloClient() }

  handleSignin = (token) => {
    localStorage.setItem('token', token);
    this.setState(() => ({ loggedIn: true }));
  }

  handleSignout = () => {
    localStorage.removeItem('token');
    const { client } = this.state;
    client.resetStore();
    this.setState(() => ({ loggedIn: false }));
  }

  render() {
    const { loggedIn, client } = this.state;
    return (
      <ApolloProvider client={client}>
        <Router>
          <div className="App">
            <Navigation
              loggedIn={loggedIn}
              signout={this.handleSignout}
            />
            <Routes loggedIn={loggedIn} signin={this.handleSignin} />
          </div>
        </Router>
      </ApolloProvider>
    );
  }
}

export default App;
