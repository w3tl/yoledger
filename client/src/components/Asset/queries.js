import gql from 'graphql-tag';
import fragments from './fragments';

export const LIST_QUERY = gql`
query AssetsQuery {
  accounts(type: ASSET) {
    ...AssetFormAccount
  }
}
${fragments.asset}
`;


export const FORM_QUERY = gql`
query AssetQuery($id: ID!) {
  account(id: $id) {
    ...AssetFormAccount
  }
}
${fragments.asset}
`;

export const ADD_MUTATION = gql`
mutation addAccount($input: AddAccountInput!) {
  addAccount(input: $input) {
    account {
      ...AssetFormAccount
    }
  }
}
${fragments.asset}
`;
