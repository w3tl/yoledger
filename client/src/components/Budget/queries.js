import gql from 'graphql-tag';

export const BODY_QUERY = gql`
query BudgetPeriodsQuery($dateStart: Date!, $count: Int) {
  budgets(dateStart: $dateStart, count: $count) {
    periods
    accounts {
      name
    }
  }
}
`;

export const COLUMN_QUERY = gql`
query BudgetQuery($date: Date!) {
  budget(date: $date) {
    id
    account {
      name
    }
    amount
  }
}
`;

export const UPDATE_MUTATION = gql`
mutation upsertBudget($input: UpsertBudgetInput!) {
  upsertBudget(input: $input) {
    success
    allocation {
      id
      account { id name }
      amount
    }
  }
}
`;
