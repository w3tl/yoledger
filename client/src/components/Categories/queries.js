import gql from 'graphql-tag';
import fragments from './fragments';

export const LIST_QUERY = gql`
query CategoriesQuery {
  expenses: accounts(type: EXPENSE) {
    ...Category
  }
  incomes: accounts(type: INCOME) {
    ...Category
  }
}
${fragments.category}
`;

export const ADD_MUTATION = gql`
mutation addAccount($input: AddAccountInput!) {
  addAccount(input: $input) {
    account {
      ...Category
    }
  }
}
${fragments.category}
`;
