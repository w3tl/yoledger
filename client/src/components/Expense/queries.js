import gql from 'graphql-tag';
import fragments from './fragments';

export const LIST_QUERY = gql`
query ExpensesQuery {
  accounts(type: EXPENSE) {
    ...ExpenseAccount
  }
}
${fragments.expense}
`;

export const FORM_QUERY = gql`
query ExpenseQuery($id: ID!) {
  account(id: $id) {
    ...ExpenseAccount
  }
}
${fragments.expense}
`;

export const ADD_MUTATION = gql`
mutation addAccount($input: AddAccountInput!) {
  addAccount(input: $input) {
    account {
      ...ExpenseAccount
    }
  }
}
${fragments.expense}
`;
