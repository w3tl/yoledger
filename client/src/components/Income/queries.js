import gql from 'graphql-tag';
import fragments from './fragments';

export const LIST_QUERY = gql`
query IncomesQuery {
  accounts(type: INCOME) {
    ...IncomeAccount
  }
}
${fragments.income}
`;

export const FORM_QUERY = gql`
query IncomeQuery($id: ID!) {
  account(id: $id) {
    ...IncomeAccount
  }
}
${fragments.income}
`;

export const ADD_MUTATION = gql`
mutation addAccount($input: AddAccountInput!) {
  addAccount(input: $input) {
    account {
      ...IncomeAccount
    }
  }
}
${fragments.income}
`;
