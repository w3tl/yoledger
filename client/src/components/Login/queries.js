import gql from 'graphql-tag';

export const SIGNIN_MUTATION = gql`
mutation signin($login: String!, $password: String!) {
  signin(login: $login, password: $password) {
    token
  }
}
`;

export const SIGNOUT_MUTATION = gql`
mutation signout($token: String) {
  signout(token: $token) {
    success
  }
}
`;
