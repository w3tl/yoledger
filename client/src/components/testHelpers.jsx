import React from 'react';
import { shallow, mount } from 'enzyme';
import { MockedProvider } from 'react-apollo/test-utils';

export const wait = (ms = 0) => new Promise(resolve => setTimeout(resolve, ms));

export const withProvider = Component => props => (
  <MockedProvider {...props}>
    <Component />
  </MockedProvider>
);

export const testRenderWithoutError = (Component) => {
  it('renders without error', async () => {
    const wrapper = shallow(Component);
    await wait();
    wrapper.update();
    expect(wrapper).toMatchSnapshot();
  });
};

export const testLoadingState = (Component) => {
  it('should render loading state initially', () => {
    const wrapper = mount(Component);
    expect(wrapper.text()).toEqual('Loading...');
  });
};

export const testErrorUI = (Component) => {
  it('should show error UI', async () => {
    const wrapper = mount(Component);
    await wait(0); // wait for response
    expect(wrapper.update().text()).toContain('Error');
  });
};

// THNX: http://blog.blakesimpson.co.uk/read/98-how-to-stub-date-in-jest-tests-building-a-stubdate-helper
export const stubDate = (fixedDate) => {
  let originalDate;

  beforeAll(() => {
    originalDate = Date;

    global.Date = class extends Date {
      constructor() {
        super();

        return fixedDate;
      }
    };
  });

  afterAll(() => {
    global.Date = originalDate;
  });
};
