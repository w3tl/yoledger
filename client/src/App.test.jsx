/* eslint-disable react/prop-types */
import React from 'react';
import ReactDOM from 'react-dom';
import { mount } from 'enzyme';
import { wait, client, stubDate } from './components/testHelpers/index';
import App from './App';
import { GET_PAGE_SETTINGS } from './components/Transaction/queries';

it('renders without crashing', () => {
  const div = document.createElement('div');
  ReactDOM.render(<App />, div);
  ReactDOM.unmountComponentAtNode(div);
});

jest.mock('./components/Routes', () => ({ signin }) => (
  <button id="signin" type="submit" onClick={() => signin('new token')} />
));

describe('App', () => {
  stubDate('2017-12-31');

  it('should unset token and clear store after sign out', async () => {
    localStorage.setItem('token', 'eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJleHAiOjE1MTQ3NjQ4MDB9.fbvqGzT4kOW9MBvNux5QHiSv479qx3-wiVuAiCUkBSc');
    client.writeQuery({
      query: GET_PAGE_SETTINGS,
      variables: { type: 'INCOME' },
      data: { settings: { __typename: 'AppSettings', transaction: { __typename: 'TransactionSettings', date: '2018-01-01', period: 'DAY' } } },
    });
    const output = mount(<App />);
    await wait();
    expect(Object.keys(client.store.cache.data.data).length).toBeGreaterThan(0);
    output.find('[name="signout"]').simulate('click');
    await wait(10);
    expect(localStorage.getItem('token')).toBeNull();
    expect(client.store.cache.data.data.ROOT_QUERY).toMatchSnapshot();
  });

  it('should set token after sign in', async () => {
    localStorage.removeItem('token');
    const output = mount(<App />);
    await wait();
    output.find('#signin').simulate('click');
    await wait(5);
    expect(localStorage.getItem('token')).toBeDefined();
  });
});

// expect(Object.keys(client.store.cache.data.data).length).toBeGreaterThan(0);

// client.writeQuery({
//   query: GET_PAGE_SETTINGS,
//   variables: { type: 'INCOME' },
//   data: { settings: { __typename: 'AppSettings', transaction: { __typename: 'TransactionSettings', date: '2018-01-01', period: 'DAY' } } },
// });
// expect(localStorage.getItem('token')).toBeNull();
// expect(Object.keys(client.store.cache.data.data)).toHaveLength(0);
