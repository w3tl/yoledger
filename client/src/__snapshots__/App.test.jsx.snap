// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App should unset token and clear store after sign out 1`] = `
Object {
  "settings": Object {
    "generated": true,
    "id": "$ROOT_QUERY.settings",
    "type": "id",
    "typename": "AppSettings",
  },
}
`;
